src/debot/term_encryption_box.rs

```diff
@@ -0,0 +1,240 @@
+use base64;
+use std::io::{self};
+use ton_client::crypto::{
+    chacha20, nacl_box, nacl_box_open, nacl_secret_box, nacl_secret_box_open,
+    register_encryption_box, EncryptionBoxHandle, EncryptionBoxInfo, ParamsOfChaCha20,
+    ParamsOfNaclBox, ParamsOfNaclBoxOpen, ParamsOfNaclSecretBox, ParamsOfNaclSecretBoxOpen,
+};
+use ton_client::error::ClientResult;
+
+#[derive(Clone, Copy)]
+pub(crate) enum EncryptionBoxType {
+    SecretNaCl,
+    NaCl,
+    ChaCha20,
+}
+
+pub(crate) struct ParamsOfTerminalEncryptionBox {
+    pub box_type: EncryptionBoxType,
+    pub their_pubkey: String,
+    pub nonce: String,
+    pub context: TonClient,
+}
+
+pub struct NaClSecretBox {
+    /// 256-bit key - unprefixed 0-padded to 64 symbols hex string.
+    pub key: String,
+    /// 96-bit nonce, encoded in `hex`.
+    pub nonce: String,
+    /// Client params.context.
+    pub client: TonClient,
+}
+
+pub struct ChaChaBox {
+    /// 256-bit key, encoded with `base64`.
+    pub key: String,
+    /// 96-bit nonce, encoded in `hex`.
+    pub nonce: String,
+    /// Client params.context.
+    pub client: TonClient,
+}
+
+pub struct NaClBox {
+    /// Receiver's public key - unprefixed 0-padded to 64 symbols hex string.
+    pub their_pubkey: String,
+    /// Sender's private key - unprefixed 0-padded to 64 symbols hex string.
+    pub secret: String,
+    /// Nonce, encoded in `hex`.
+    pub nonce: String,
+    /// Client params.context.
+    pub client: TonClient,
+}
+
+#[async_trait::async_trait]
+impl ton_client::crypto::EncryptionBox for NaClSecretBox {
+    async fn get_info(&self) -> ClientResult<EncryptionBoxInfo> {
+        Ok(EncryptionBoxInfo {
+            hdpath: Some(String::from("0")),
+            algorithm: Some("NaclSecretBox".to_string()),
+            options: Some(json!({"nonce": self.nonce.clone()})),
+            public: None,
+        })
+    }
+    async fn encrypt(&self, data: &String) -> ClientResult<String> {
+        Ok(nacl_secret_box(
+            self.client.clone(),
+            ParamsOfNaclSecretBox {
+                decrypted: base64::encode(data),
+                key: self.key.clone(),
+                nonce: self.nonce.clone(),
+            },
+        )
+        .unwrap()
+        .encrypted)
+    }
+    async fn decrypt(&self, data: &String) -> ClientResult<String> {
+        Ok(nacl_secret_box_open(
+            self.client.clone(),
+            ParamsOfNaclSecretBoxOpen {
+                encrypted: data.clone(),
+                key: self.key.clone(),
+                nonce: self.nonce.clone(),
+            },
+        )
+        .unwrap()
+        .decrypted)
+    }
+}
+
+#[async_trait::async_trait]
+impl ton_client::crypto::EncryptionBox for ChaChaBox {
+    async fn get_info(&self) -> ClientResult<EncryptionBoxInfo> {
+        Ok(EncryptionBoxInfo {
+            hdpath: Some(String::from("0")),
+            algorithm: Some("ChaCha20".to_string()),
+            options: Some(json!({"nonce": self.nonce.clone()})),
```

Comment:
It's very important that the structure of `options` param is the same as in Surf Browser. I think we must standardize it in the consortium.